pkg/runtime/namespace/namespace.go

```diff
@@ -164,6 +164,49 @@ func (ns *Namespace) DB(ctx context.Context, group string) proto.DB {
 	return exist[target]
 }
 
+// DBMaster returns a master DB, returns nil if nothing selected.
+func (ns *Namespace) DBMaster(_ context.Context, group string) proto.DB {
+	// use weight manager to select datasource
+	dss := ns.dss.Load().(map[string][]proto.DB)
+	exist, ok := dss[group]
+	if !ok {
+		return nil
+	}
+	// master weight w>0 && r>0
+	for _, db := range exist {
+		if db.Weight().W > 0 && db.Weight().R > 0 {
+			return db
+		}
+	}
+	return nil
+}
+
+// DBSlave returns a slave DB, returns nil if nothing selected.
+func (ns *Namespace) DBSlave(_ context.Context, group string) proto.DB {
+	// use weight manager to select datasource
+	dss := ns.dss.Load().(map[string][]proto.DB)
+	exist, ok := dss[group]
+	if !ok {
+		return nil
+	}
+	var (
+		target     = 0
+		wrList     = make([]int, 0, len(exist))
+		readDBList = make([]proto.DB, 0, len(exist))
+	)
+	// slave weight w==0 && r>0
+	for _, db := range exist {
+		if db.Weight().W == 0 && db.Weight().R > 0 {
+			wrList = append(wrList, int(db.Weight().R))
+			readDBList = append(readDBList, db)
+		}
```

Comment:
My mistake, please change the slave logic to `W==0 && R==0`.

Reply:
> My mistake, please change the slave logic to `W==0 && R==0`.

why the slave node  `R==0`?

Reply:
The final conclusion:
- MASTER: w > 0 && r >= 0
- SLAVE: w == 0 && r >= 0


PS: the `r==0` has high priority.

Reply:
done